### PR TITLE
add component/foreach helpers

### PR DESCRIPTION
**Pull request by Stebalien**
This adds a `Component` helper type and a `ForEach` helper method.

The first attempt used an interface but interfaces imply allocation. We really
can't afford to allocate here.

### CODE REVIEW

**Comment by Stebalien** [comment]
Note: this changes the behavior of paths. `ValueForProtocol(pathProtocol)` will now make the path start with a `/`.

Rational:

1. Paths should start with `/`.
2. This makes the string-to-bytes and bytes-to-string encoders symmetric. In the past, the former used to take a path starting with a `/` and the latter would spit out a path without the `/`.

File: component.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+func (c *Component) Equal(o Multiaddr) bool {
+	return bytes.Equal(c.bytes, o.Bytes())
+}
+
+func (c *Component) Protocols() []Protocol {
+	return []Protocol{c.protocol}
+}
+
+func (c *Component) Decapsulate(o Multiaddr) Multiaddr {
```
Comment (raulk) [inline component.go]: I guess `Decapsulate` doesn't make much sense in the context of a single component, so it becomes equivalent to a Filter operation, i.e. returns itself if there's a match?

File: component.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		bytes:    maddr,
+		protocol: protocol,
+		offset:   offset,
+	}
+}
+
+// ForEach walks over the multiaddr, component by component.
+//
+// This function iterates over components *by value* to avoid allocating.
+func ForEach(m Multiaddr, cb func(c Component) bool) {
```
Comment (raulk) [inline component.go]: Would it make sense to add a `Filter` function that returns the components for a given protocol?

Comment (Stebalien) [reply component.go]: Not sure about filter but I'm adding `SplitFirst`, `SplitLast`, `SplitFunc`, etc. functions.

File: component.go (hunk 1 above)
Comment (Stebalien) [reply component.go]: Basically. Personally, given the new `SplitFunc`, I'm not sure how useful decapsulate is in general.

**Pull request merged**